scipy/stats/stats.py

```diff
@@ -2614,10 +2592,36 @@ def zmap(scores, compare, axis=0, ddof=0):
     array([-1.06066017,  0.        ,  0.35355339,  0.70710678])
 
     """
-    scores, compare = map(np.asanyarray, [scores, compare])
-    mns = compare.mean(axis=axis, keepdims=True)
-    sstd = compare.std(axis=axis, ddof=ddof, keepdims=True)
-    return (scores - mns) / sstd
+    a = np.asanyarray(compare)
+
+    if a.size == 0:
+        return np.empty(a.shape)
+
+    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+
+    if contains_nan and nan_policy == 'omit':
+        if axis is None:
+            mn = _quiet_nanmean(a.ravel())
+            std = _quiet_nanstd(a.ravel())
+            isconst = _isconst(a.ravel())
+        else:
+            mn = np.apply_along_axis(_quiet_nanmean, axis, a)
+            std = np.apply_along_axis(_quiet_nanstd, axis, a)
```

Comment:
Why don't we need to `keepdims` here?

Reply:
`_quiet_nanmean` and `_quiet_nanstd` will take care of `keepdims=True`.

We use `keepdims=True` when computing the mean and std. dev. so that broadcasting works correctly when we calculate `z = (scores - mn) / std`.  `mn` and `std` must maintain shapes that are compatible for broadcasting with `scores`.  If, for example, `scores` has shape `(3, 4, 5)` and we use `axis=1`, we want `mn` and `std` to have shape `(3, 1, 4)`.  If we didn't use `keepdims`, they would have shape `(3, 4)` and couldn't broadcast with `scores`.